YARP overview page (#34764)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/servers/yarp/yarp-overview.md

## What a reverse proxy does

A reverse proxy is a server that sits between client devices and backend servers. It forwards client requests to the appropriate backend server and then returns the server's response to the client. A reverse proxy provides several benefits:

- Routing: Directs requests to different backend servers based on predefined rules, such as URL patterns or request headers. For example, `/images`, `/api`, `/db` requests can be can be routed image, api, and database severs.
- Load Balancing: Distributes incoming traffic across multiple backend servers to prevent overloading a specific server. Distribution increases performance and reliability.
- [SSL/TSL Termination](/azure/application-gateway/ssl-overview): Offloads the TSL encryption and decryption processes from backend servers, reducing their workload.
- Connection abstraction, decoupling and control over URL-space: Inbound requests from external clients and outbound responses from the backend are independent. This independence allows for differnt:
  - Versions of HTTP, ie, HTTP/1.1, HTTP/2, HTTP/3. The proxy can upgrade or downgrade HTTP versions.

>   - Connection lifetimes, which enables features like long-polling on the backend while maintaining short client connections.

Reviewer: ```suggestion
  - Connection lifetimes, which enables having long-lived connections on the backend while maintaining short client connections.
```

  - Control Over URL-Space: Incoming URLs can be transformed before forwarding to the backend. This abstracts the external URLs from how they are mapped to internal services. Internal service endpoints can change without affecting external URLs.

> - Security: Internal service endpoints can be hidden from external exposure, protecting against some types of cyber attacks such as [DDoS attacks](https://www.microsoft.com/security/business/security-101/what-is-a-ddos-attack?msockid=3e35ed3aa4666d8003aaf830a5006c74) and [SQL injection ](/sql/relational-databases/security/sql-injection).

Reviewer: How do we help with SQL injection?

> - Security: Internal service endpoints can be hidden from external exposure, protecting against some types of cyber attacks such as [DDoS attacks](https://www.microsoft.com/security/business/security-101/what-is-a-ddos-attack?msockid=3e35ed3aa4666d8003aaf830a5006c74) and [SQL injection ](/sql/relational-databases/security/sql-injection).

Reviewer 2: CoPilot hallucinations, I should have checked.
```suggestion
- Security: Internal service endpoints can be hidden from external exposure, protecting against some types of cyber attacks such as [DDoS attacks](https://www.microsoft.com/security/business/security-101/what-is-a-ddos-attack?msockid=3e35ed3aa4666d8003aaf830a5006c74).
```

- Caching: Frequently requested resources can be cached to reduce the load on backend servers and improve response times.
- Versioning: Different versions of an API can be supported using different URL mappings.

## Why use a reverse proxy

> Using a reverse proxy offers several advantages:

Reviewer: This section feels redundant with the "What a reverse proxy does". I'd merge the two if possible.

Reviewer 2: I agree, CoPilot and myself duplicated items.
@samsp-msft I can do that if you like.

@samsp-msft  originally had the 2 sections, albeit with less duplicates.

- Simplified maintenance: Reverse proxies can handle [SSL/TSL Termination](/azure/application-gateway/ssl-overview) and other tasks, simplifying the configuration and maintenance of backend servers. For example, SSL certificates and security policies can be managed at the reverse proxy level instead of on each individual server.

## How a reverse proxy handles HTTP

A reverse proxy handles HTTP requests and responses in the following manner:

- Receiving  requests: The reverse proxy listens on specified ports and endpoints for incoming HTTP requests from clients.
- Terminating Connections: The inbound HTTP connections are terminated at the proxy and new connections are used for outbound requests to destinations.
- Routing  requests: Based on predefined routing rules and configurations, the reverse proxy determines which backend server, or cluster of servers, should handle the request.
- Forwarding  requests: The reverse proxy forwards the client request to the appropriate backend server, transforming the path and headers as necessary.

> - Connection pooling: The outbound connections are pooled to reduce connection overhead and make most use of HTTP 1.1 pipelining and parallel requests with HTTP/2 and HTTP/3.

Reviewer: ```suggestion
- Connection pooling: The outbound connections are pooled to reduce connection overhead and make most use of HTTP 1.1 reuse and parallel requests with HTTP/2 and HTTP/3.
```
HTTP/1.1 pipelining is a feature we're intentionally not supporting at the HttpClient layer.
